### PR TITLE
Make fused grouped-expert projection the default

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4616
* #4614
* #4612
* #4611
* #4550
* __->__ #4552
* #4526
* #4546

### CODE REVIEW

File: torchtitan/models/kimi_k3/moe.py [hunk 1]
```diff
@@ -67,42 +66,14 @@ def __init__(self, config: Config):
         self.beta = config.beta
         self.linear_beta = config.linear_beta
 
-    def forward(
+    def _activation(
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/moe.py]: commented in the linear PR

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -37,25 +39,104 @@
...
+    del state_shardings["w1_EFD"]
+    del state_shardings["w3_EFD"]
+    state_shardings["w13"] = gate_sharding
+    return replace(sharding_config, state_shardings=state_shardings)
+
+
 class GroupedExperts(Module):
+    """SwiGLU experts with one physical interleaved gate-up parameter.
+
+    ``w13`` has shape ``(E, F, 2, D)`` and is flattened to ``(E, 2F, D)`` for
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: why don't we define w13 as (E, 2F, D) as in e.g. `FusedQKV` https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/config_utils.py#L244

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -37,25 +39,104 @@
...
+
+    def init(param_EF2D: torch.Tensor) -> None:
+        with torch.no_grad():
+            gate_init(param_EF2D[:, :, 0, :])
+            up_init(param_EF2D[:, :, 1, :])
+
+    return init
+
+
+def _fuse_grouped_experts_param_init(param_init: dict | None) -> dict | None:
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: can be put together with https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/config_utils.py#L109

In some sense, this is not part of core torchtitan. Users can do whatever.

**Comment by anijain2305** [comment]
Talked to @tianyu-l - we will wait for @jinsooihm  to make GroupedExpert to only contain one weight.

Update - We will land this first, and then @jinsooihm will refactor after it.

File: torchtitan/overrides/fused_swiglu.py [hunk 4]
```diff
@@ -401,128 +399,28 @@ def dist_gemm_fused_swiglu(
...
-        del self.w3_EFD
-        self.w13 = torch.nn.Parameter(
-            torch.empty(config.num_experts, config.hidden_dim, 2, config.dim)
-        )
-
-        self.register_state_dict_post_hook(self._split_w13_on_save)
-        self.register_load_state_dict_pre_hook(self._merge_w13_on_load)
-
-    def forward(
+    def _activation(
```
Comment (tianyu-l) [inline torchtitan/overrides/fused_swiglu.py]: related to my comment on first PR:

If we remove this, it seems the override only needs to happen on `ActivationFn`, not `GroupedExperts` which is nice, because we no longer need to separately override MLP and GroupedExperts

File: torchtitan/models/common/moe.py [hunk 5]
```diff
@@ -58,19 +91,29 @@ class Config(Module.Config):
             )
         )
 
+        def build(self, **kwargs):
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: Can this happen only in init? like in https://github.com/pytorch/torchtitan/blob/main/torchtitan/overrides/fused_swiglu.py#L621

Also, since we are making this default, we don't need to keep unfused sharding anyway. This is somewhat related to "we don't need to keep state_dict unfused" but I prefer we should do it in this PR, to avoid introducing code like `_fuse_grouped_experts_sharding`, although my opinion is not that strong.

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -85,18 +128,15 @@ def forward(
         local token shard. Keep logical capital suffixes here to avoid encoding
         a specific parallel layout in these local tensor names.
         """
-        if isinstance(self.w1_EFD, DTensor):
+        if isinstance(self.w13, DTensor):
             # Convert parameters from DTensors to plain Tensors, to work with
             # dynamic-shape inputs in EP which cannot be easily expressed as DTensors.
-            w1_EFD = self.w1_EFD.to_local()
+            w13_E2FD = self.w13.to_local()
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: technically this notation is misleading, because each letter represents a single dim

Here's previous "workaround" https://github.com/pytorch/torchtitan/blob/2d04249c028132c951b2c9a8d62f465a9b7cc1dd/torchtitan/overrides/fused_swiglu.py#L651

File: torchtitan/models/common/activation.py [hunk 7]
```diff
@@ -22,7 +22,13 @@ class Config(Configurable.Config):  # pyrefly: ignore[bad-override]
         pass
 
     @abstractmethod
-    def __call__(self, gate: torch.Tensor, up: torch.Tensor) -> torch.Tensor:
+    def __call__(
+        self,
+        gate: torch.Tensor,
+        up: torch.Tensor,
+        *,
+        offsets: torch.Tensor | None = None,
```
Comment (anijain2305) [inline torchtitan/models/common/activation.py]: to alllow offset aware activations - silu_and_mul_op - for MoE.

Comment (tianyu-l) [inline torchtitan/models/common/activation.py]: since this is generic, we don't need to have boilerplate here -- maybe make it `**kwargs`

File: torchtitan/models/common/feed_forward.py [hunk 8]
```diff
@@ -48,48 +47,100 @@ def _init(t: torch.Tensor) -> None:
...
+    for key in tuple(state_dict):
+        if not key.endswith("w13"):
+            continue
+        prefix = key.removesuffix("w13")
+        gate_up = state_dict.pop(key)
+        state_dict[f"{prefix}w1_EFD"] = gate_up[:, :, 0, :].contiguous()
+        state_dict[f"{prefix}w3_EFD"] = gate_up[:, :, 1, :].contiguous()
+
+
+def split_fused_gate_up_state_dict(
```
Comment (tianyu-l) [inline torchtitan/models/common/feed_forward.py]: if they are only used by HF state dict adapters, they probably should be put in https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/utils.py

File: torchtitan/models/common/feed_forward.py [hunk 9]
```diff
@@ -48,48 +47,100 @@ def _init(t: torch.Tensor) -> None:
     return _init
 
 
-def _make_fused_linear_init(gate_init: Callable, up_init: Callable) -> Callable:
-    """Build an initializer for an interleaved 2D gate/up linear weight."""
-    init_logical_weight = _make_fused_gate_up_init(gate_init, up_init, gate_up_axis=1)
+def _make_fused_linear_init(
```
Comment (tianyu-l) [inline torchtitan/models/common/feed_forward.py]: I feel this doesn't belong to core definition of FeedForward, so let's put together with https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/config_utils.py#L110

File: torchtitan/models/common/moe.py [hunk 10]
```diff
@@ -81,18 +92,15 @@ def forward(
         local token shard. Keep logical capital suffixes here to avoid encoding
         a specific parallel layout in these local tensor names.
         """
-        if isinstance(self.w1_EFD, DTensor):
+        if isinstance(self.w13, DTensor):
             # Convert parameters from DTensors to plain Tensors, to work with
             # dynamic-shape inputs in EP which cannot be easily expressed as DTensors.
-            w1_EFD = self.w1_EFD.to_local()
+            w13_E_F_2_D = self.w13.to_local()
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: EF2D is proper? as there are indeed 4 dims

File: torchtitan/models/common/moe.py [hunk 11]
```diff
@@ -107,9 +115,15 @@ def forward(
                 # TODO(pianpwk): likely relax this in spmd_types.
                 spmd.mutate_type(offsets_E, axis, src=spmd.P, dst=spmd.V)
 
-        gate_RF = self._grouped_mm(A=x_RD.bfloat16(), weight_EOI=w1_EFD, offs=offsets_E)
-        up_RF = self._grouped_mm(A=x_RD.bfloat16(), weight_EOI=w3_EFD, offs=offsets_E)
-        h_RF = self.activation_fn(gate_RF, up_RF)
+        E, F, _, D = w13_E_F_2_D.shape
+        w13_E_2F_D = w13_E_F_2_D.reshape(E, F * 2, D)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: in comparison, E_2F_D is a reasonable workaround

File: torchtitan/models/deepseek_v3/sharding.py [hunk 12]
```diff
@@ -33,11 +33,10 @@
     )
 
 
-# Routed-expert layout for the shared ``GroupedExperts`` (w1/w2/w3).
+# Routed-expert layout for the shared ``GroupedExperts`` (w13/w2).
 _GROUPED_EXPERTS_PARAM_LAYOUT: dict[str, spmd.PerMeshAxisSpmdType] = {
-    "w1_EFD": spmd.S(1),
+    "w13": spmd.S(1),
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/sharding.py]: `w13_EF2D`

File: torchtitan/models/deepseek_v4/sharding.py [hunk 13]
```diff
@@ -41,9 +41,8 @@
     )
 
 _GROUPED_EXPERTS_PARAM_LAYOUT: dict[str, spmd.PerMeshAxisSpmdType] = {
-    "w1_EFD": spmd.S(1),
+    "w13": spmd.S(1),
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v4/sharding.py]: similar

File: torchtitan/models/kimi_k3/sharding.py [hunk 14]
```diff
@@ -105,9 +105,8 @@ def set_kimi_k3_sharding_config(
                 # without sequence parallel.
                 enable_sp=enable_sp,
                 expert_param_layout={
-                    "w1_EFD": spmd.S(1),
+                    "w13": spmd.S(1),
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/sharding.py]: and other files

**Comment by anijain2305** [comment]
Will need to revisit this after https://github.com/pytorch/torchtitan/pull/4535 lands specifically about the (E, F, 2, D) vs (E, 2F, D) layout. If we can make the two consistent, it will be better for the readability.

Update - The PR now has  makes the grouped experts have the same layout of (E, 2F, D) for `w13` as we have feedforward `w13`.

File: torchtitan/models/common/moe.py [hunk 15]
```diff
@@ -88,13 +119,14 @@ def forward(
...
-        up_RF = self._grouped_mm(
-            A=x_RD.bfloat16(), weight_EOI=self.w3_EFD, offs=offsets_E
+        F = self.w13_E_2F_D.shape[1] // 2
+        gate_up_R_2F = self._grouped_mm(
+            A=x_RD.bfloat16(),
+            weight_EOI=self.w13_E_2F_D,
+            offs=offsets_E,
         )
-        h_RF = self.activation_fn(gate_RF, up_RF)
+        gate_RF, up_RF = gate_up_R_2F.reshape(-1, F, 2).unbind(-1)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:128]: are you saying this unbind(-1) is pretty cheap?

File: torchtitan/models/common/moe.py
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:1]: Is this PR bitwise numerics preserving?

**Review by tianyu-l — COMMENTED** [review]
This version looks very nice.

If @sanketpurandare and we agree with the layout, I'll stamp. O/w I need to take another look on the (E, 2, F, D) version. Thanks!

**Comment by anijain2305** [comment]
@sanketpurandare will take over the Pr
